nncf/experimental/torch/nas/bootstrapNAS/search/search.py

```diff
@@ -0,0 +1,492 @@
+        torch.save(evaluator_handlers_state, Path(self.checkpoint_save_dir, 'evaluators_state.pth'))
+
+    def evaluators_to_csv(self) -> NoReturn:
+        """
+        Export evaluators' information used by search algorithm to CSV
+        :return:
+        """
+        for evaluator_handler in self.evaluator_handlers:
+            evaluator_handler.export_cache_to_csv(self._log_dir)
+
+    @property
+    def efficiency_evaluator_handler(self):
+        return self._efficiency_evaluator_handler
+
+    @property
+    def accuracy_evaluator_handler(self):
+        return self._accuracy_evaluator_handler
+
+
+class SearchProblem(Problem):
+    """
+    Pymoo problem with design variables and evaluation methods.
+    """
+
+    def __init__(self, search: SearchAlgorithm):
+        """
+        Initializes search problem
+
+        :param search: search algorithm.
+        """
+        super().__init__(n_var=search.num_vars,
+                         n_obj=search.num_obj,
+                         n_constr=search.search_params.num_constraints,
+                         xl=search.vars_lower,
+                         xu=search.vars_upper,
+                         type_var=search.type_var)
+        self._search = search
+        self._search_records = search.search_records
+        self._elasticity_handler = self._search._elasticity_ctrl.multi_elasticity_handler
+        self._dims_enabled = self._elasticity_handler.get_available_elasticity_dims()
+        self._iter = 0
+        self._evaluator_handlers = search.evaluator_handlers
+        self._accuracy_evaluator_handler = search.accuracy_evaluator_handler
+        self._efficiency_evaluator_handler = search.efficiency_evaluator_handler
+        self._model = search._model
+        self._lower_bound_acc = search.search_params.ref_acc - search.acc_delta
+
+    def _evaluate(self, x: List[float], out: Dict[str, Any], *args, **kargs) -> NoReturn:
+        """
+        Evaluates a population of sub-networks.
+
+        :param x: set of sub-networks to evaluate.
+        :param out: measurements obtained by evaluating sub-networks.
+        :param args:
+        :param kargs:
+        :return:
+        """
+        evaluators_arr = [[] for i in range(len(self._search.evaluator_handlers))]
+
+        for _, x_i in enumerate(x):
+            sample = self._elasticity_handler.get_config_from_pymoo(x_i)
+            self._elasticity_handler.activate_subnet_for_config(sample)
+            if sample != self._elasticity_handler.get_active_config():
+                nncf_logger.warning("Requested configuration was invalid")
+                nncf_logger.warning("Requested: {sample}".format(sample=sample))
+                nncf_logger.warning("Provided: {config}".format(
+                                    config=self._elasticity_handler.get_active_config()))
+                self._search.bad_requests.append((sample, self._elasticity_handler.get_active_config()))
```

Comment:
For elastic depth we can directly sample only valid configuration in terms of progressivity and depth_indicator.
1) choose a list of groups being selected
2) choose how many layers are skipped for each group. 
It should unambiguously define each valid configuration and reduce redundant logic with deleting blocks not satisfying progressivity.

Reply:
As we discussed today, these bad requests are related to the multi-dimensional validation in multielasticity handler. The Search component receives a filtered depth search space, so this change won't affect the search component. Having said that, I really like the idea of simplifying the generation of the depth search spaces with the indexes idea. :) It will be faster to generate the depth search space. Thanks!

Reply:
create a follow-up ticket (83708) for that improvement. as discussed internally, shouldn't affect search, it only affects sampling during supernet training